Create JEP-231: Remove commons-digester dependency from Jenkins Core (#361)
jenkinsci/jep: jep/0000/README.adoc

> = JEP-0000: Remove commons-digester dependency from Jenkin Core

Reviewer: ```suggestion
= JEP-0000: Remove commons-digester dependency from Jenkins Core
```

:toc: preamble
:toclevels: 3
ifdef::env-github[]
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
endif::[]

.Metadata
[cols="1h,1"]
|===
| JEP
| 0000

| Title

> | Remove commons-digester dependency from Jenkin Core

Reviewer: ```suggestion
| Remove commons-digester dependency from Jenkins Core
```

| Sponsor
| link:https://github.com/olamy[Olivier Lamy], link:https://github.com/bitwiseman[Liam Newman]

// Use the script `set-jep-status <jep-number> <status>` to update the status.
| Status
| Not Submitted :information_source:

| Type
| Informational

| Created
| (2021-05-07)

//
//
// Uncomment if there is an associated placeholder JIRA issue.
//| JIRA
//| :bulb: https://issues.jenkins-ci.org/browse/JENKINS-nnnnn[JENKINS-nnnnn] :bulb:
//
//
// Uncomment if discussion will occur in forum other than jenkinsci-dev@ mailing list.
//| Discussions-To

> //| :bulb: Link to where discussion and final status announcement will occur :bulb:

Reviewer: I believe we had this discussion a few years ago. If we do not link the discussion, it will be hard for visitors to find the right thread

|===

== Abstract

This JEP describes the steps taken to remove link:https://commons.apache.org/proper/commons-digester/[Apache Commons Digester] as a dependency of Jenkins Core.

== Specification

Removing link:https://commons.apache.org/proper/commons-digester/[Apache Commons Digester] from Jenkins Core requires changes to Core and to a number of plugins.
We will describe these two groups of changes separately.

=== Core

Jenkins Core implemented 
link:https://github.com/jenkinsci/jenkins/blob/22fcc54bd3879c765b5595371e57fd180860daff/core/src/main/java/hudson/util/Digester2.java[`hudson.util.Digester2`] 
which extended `org.apache.commons.digester.Digester`
attempting to set secure parsing defaults to prevent XXE (XML External Entity) vulnerabilities.
Neither class was used by Jenkins Core, `hudson.util.Digester2` was purely provided as a helper for plugins.

=== Plugins

Plugins will need to be updated to avoid breaking during this change.
The changes needed are relatively straightforward.
Assuming the dependency cannot be removed, the code changes needed are generally not complex.

==== Example change to plugin code

> Plugins can be updated using the following code snippet: 

Author: @jglick 
Is the "Insert a suggestion" button not enabled for you?  
Okay.
```suggestion
Plugins may be better off not using Digester at all, taking this opportunity to replace some scary-looking reflection with plain old Java programming.
The changes needed in that case will vary depending on the plugin, but an example can be seen in
link:https://github.com/jenkinsci/mercurial-plugin/commit/84af58b08f80bb92792f7bc04a31487f3eeee95a[ jenkinsci/mercurial-plugin @ `84af58b`]

If that is too involved, plugins can also be updated using the following code snippet to configure Digester with more secure settings: 
```

[source,java]
----
import org.apache.commons.digester3.Digester;
....
public static Digester createDigester(boolean secure) throws SAXException {
    Digester digester = new Digester();
    if (secure) {
        digester.setXIncludeAware(false);
        try {
            digester.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
            digester.setFeature("http://xml.org/sax/features/external-general-entities", false);
            digester.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
            digester.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
        } catch (ParserConfigurationException ex) {
            throw new SAXException("Failed to securely configure xml digester parser", ex);
        }
    }
    return digester;
}
----

==== Effected plugins (fixes planned)

> Below is a table of the effected plugins with PRs submitted to fix their breaks.
> These PRs may or may not be merged before removal is complete, depending on whether

Reviewer 2: ```suggestion
Below is a table of the affected plugins with PRs submitted to fix their breaks.
These PRs may or may not be merged before removal is complete, depending on whether
```

maintainers merge and release the submitted PRs.

This table will be updated as needed to track the state of each plugin.

Legend:

> * :heavy_check_mark: - Fixed or no need to fix

Reviewer 2: ```suggestion
* :heavy_check_mark: Fixed or no need to fix
```

> * :stop_sign: 	     - It's not going to be fixed (no maintainer mainly)

Reviewer 2: ```suggestion
* :stop_sign: 	     It's not going to be fixed (no maintainer mainly)
```

> * :exclamation: 	 - Waiting for something (merge or release)

Reviewer 2: ```suggestion
* :exclamation: 	 Waiting for something (merge or release)
```

|  link:https://github.com/jenkinsci/synergy_scm-plugin/pull/17[jenkinsci/synergy_scm-plugin#17]) (last activity 6 years ago)
| 96

| :stop_sign: (no maintainer, needed repo missed)
| link:https://github.com/jenkinsci/config-rotator-plugin[config-rotator]
|  link:https://github.com/jenkinsci/config-rotator-plugin/pull/3[jenkinsci/config-rotator-plugin#3] (last activity 4 years ago)  :rotating_light: need help from Praqma, as https://code.praqma.net/repo/maven/ no longer exists
| 62

| :stop_sign: (no maintainer)
| link:https://github.com/jenkinsci/harvest-plugin[harvest]
|  link:https://github.com/jenkinsci/harvest-plugin/pull/5[jenkinsci/harvest-plugin#5] (last activity 6 years ago)
| 49

| :exclamation: maintainers testing
| link:https://github.com/jenkinsci/plasticscm-mergebot-plugin[plasticscm-mergebot]
|  link:https://github.com/jenkinsci/plasticscm-mergebot-plugin/pull/3[jenkinsci/plasticscm-mergebot-plugin#3]  (last active/release late 2019)
| 55

| :stop_sign: (no maintainer)
| link:https://github.com/jenkinsci/cmvc-plugin[cmvc]
|  link:https://github.com/jenkinsci/cmvc-plugin/pull/3[jenkinsci/cmvc-plugin#3]  (last activity 9 years ago...)
| 18
|===

> ==== Effected plugins (no fixes planned)

Reviewer 2: ```suggestion
==== Affected plugins (no fixes planned)
```

===== Suspended plugins

Some plugins were found which were suspended when Jenkins 2 was released.
These will not be updated as part of this effort.

* https://plugins.jenkins.io/svn-release-mgr suspended since https://issues.jenkins-ci.org/browse/INFRA-2487
* https://github.com/jenkinsci/cpptest-plugin suspended since https://issues.jenkins-ci.org/browse/INFRA-2487
* https://github.com/jenkinsci/tfs-plugin suspended since https://issues.jenkins-ci.org/browse/INFRA-2751
* https://github.com/jenkinsci/cflint-plugin (link:https://github.com/jenkinsci/CFLint-plugin/pull/3[PR-3]) suspended since https://issues.jenkins-ci.org/browse/INFRA-2751
* https://github.com/jenkinsci/script-scm-plugin SECURITY-461
* https://github.com/jenkinsci/rtc-plugin (superseded by team-concert)

===== Never released plugins

> Some plugins were found which were suspended when Jenkins 2 was released.

Reviewer 2: Duplicated:
```suggestion
Some plugins were never released.
```

It contains security vulnerabilities and also is not actually used by any part of Jenkins Core itself.
Jenkins Core exposes an API to plugin that mitigates some of these security issues, but it is not the job of Jenkins to do this.
Instead, we should remove the Digester dependency from Jenkins Core and the plugins that actually use Digester should handle these security concerns themselves.
Also, because it is a dependency of Core, some plugins have taken a dependency on Digester being provided by Jenkins rather than using the API provided by Jenkins.

Plugins in either of these scenarios will need to be updated or they will break after this removal.

== Reasoning

=== Risks of complete removal

We chose to remove Digester from Jenkins Core in a manner that causes plugins which depend on it to fail unless they are updated.
We made this choice based on an analysis of all plugins which are currently distributed by the Jenkins Update Center or which are
in a repository in the link:https://github.com/jenkinsci[`jenkinsci` org on GitHub].

> All but a few of the plugins effected by this change are some combination of unmaintained, rarely used, and no longer or never published.

Reviewer 2: ```suggestion
All but a few of the plugins affected by this change are some combination of unmaintained, rarely used, and no longer or never published.
```

> We submitted PRs to all effected plugins.

Reviewer 2: ```suggestion
We submitted PRs to all affected plugins.
```

=== Detached plugin to maintain compatibility

The current plan causes plugins which depend on Jenkins to provide Digester to fail unless they are updated.
This could be mitigated by moving this dependency to a detached plugin.

> We decided against creating a detached pluging because there were a small number effected plugins and only a few of them have significant install base.

Reviewer 3: ```suggestion
We decided against creating a detached pluging because there were a small number of effected plugins and only a few of them have significant install base.
```

> We decided against creating a detached pluging because there were a small number effected plugins and only a few of them have significant install base.

Reviewer 2: ```suggestion
We decided against creating a detached pluging because there were a small number of affected plugins and only a few of them have significant install base.
```

> The creating and maintaining a detached plugin would still be a significant amount of work and would cause the security vulnerabilities we are trying to address to remain open.

Reviewer 3: ```suggestion
The creating and maintaining of a detached plugin would still be a significant amount of work and would cause the security vulnerabilities we are trying to address to remain open.
```

== Backwards Compatibility

The current plan causes plugins which depend on Jenkins to provide Digester to fail unless they are updated.
The majority of this JEP is devoted to discussion of how to mitigate these breaks.

== Security

The motivation for this change is mitigating a number of known security issues

== Infrastructure Requirements

There are no new infrastructure requirements related to this proposal.

> == Testing

Author: I don't think so.  But I'm open to discussion on the topic.

These will not be updated as part of this effort.

* https://plugins.jenkins.io/svn-release-mgr suspended since https://issues.jenkins-ci.org/browse/INFRA-2487
* https://github.com/jenkinsci/cpptest-plugin suspended since https://issues.jenkins-ci.org/browse/INFRA-2487
* https://github.com/jenkinsci/tfs-plugin suspended since https://issues.jenkins-ci.org/browse/INFRA-2751
* https://github.com/jenkinsci/cflint-plugin (link:https://github.com/jenkinsci/CFLint-plugin/pull/3[PR-3]) suspended since https://issues.jenkins-ci.org/browse/INFRA-2751
* https://github.com/jenkinsci/script-scm-plugin SECURITY-461
* https://github.com/jenkinsci/rtc-plugin (superseded by team-concert)

===== Never released plugins

Some plugins were never released.
These will not be updated as part of this effort.

* https://github.com/jenkinsci/cocoemma-plugin
* https://github.com/jenkinsci/jwsdp-sqe-plugin
* https://github.com/jenkinsci/pucm-plugin
* https://github.com/jenkinsci/purecm-plugin

== Motivation

The link:https://commons.apache.org/proper/commons-digester/[Apache Commons Digester] that was included as a dependency of Jenkins Core is old and poorly maintained.

> It contains security vulnerabilities and also is not actually used by any part of Jenkins Core itself.

Reviewer 4: Can you include a link or set of links to CVEs here? I doubt many people know what Commons Digester even does (I originally thought it had something to do with message digests) let alone know why it's an insecure library for user input. Yes, I know that it's basically a gaping security hole that turns XML into executable code, but normal users don't know that!

Author: @jenkinsci/jep-editors 
Should I go ahead and make this a draft JEP, or should another JEP editor do this (I'm not sure I should merge my own PR.

Reviewer: Not sure as well. Is it about transitive https://snyk.io/vuln/maven:commons-beanutils:commons-beanutils . Would be nice to clarify

Reviewer 4: I don't see any problems in beanutils; it says to upgrade to 1.9.4.

Reviewer 4: And I don't have specific vulnerabilities to list other than the fact that Commons Digester is effectively an XML-based programming language, so using it to parse user input is a terrible idea.

Author: > And I don't have specific vulnerabilities to list other than the fact that Commons Digester is effectively an XML-based programming language, so using it to parse user input is a terrible idea.

That's pretty bad. 

The instigator was `beanutils` yes, which can be mitigated.

However, as noted in the next paragraph, Digester itself is easy to misconfigure. The reason `Digester2` exists is to mitigate that, but there's nothing stopping plugins from misusing it. So, why make that vulnerability easy open?

If someone with access to [SECURITY-2147](https://github.com/jenkinsci/jenkins/pull/5320/files#diff-268ec5d31daea455e70270053b53943b8471208601e34a9709e2a70c41af68cfL42) could add a link to the CVE that'd be great.

Reviewer 3: This is the one you are looking for I believe:
CVE-2020-2304

These all look related:

CVE-2020-2305
CVE-2018-1000054
CVE-2018-1000055

Probably more as well

Reviewer 5: > Digester itself is easy to misconfigure. The reason `Digester2` exists is to mitigate that

Bear in mind that there are two sorts of security issues in using Digester:

* XML is overcomplicated and the XML parsing might be subject to XXE, maybe allowing clever attacks to do stuff like load `../../../credentials.xml`.
* Even if you make sure it is being fed a plain old self-contained XML document, you are still taking user-controlled strings and using them to look up and invoke Java methods, which is subject to abuse in numerous ways.

Reviewer 4: What Jesse said is essentially the type of detail I'd love to see here.

Author: @jvz @jglick 
Awesome, please be the change you want to see and use the suggestion button to write it as you think it should be.

Reviewer 4: That may be true, but submitting a JEP that's basically "we're removing a dependency due to 'security vulnerabilities'" with no further explanation is strange. This should be a draft PR in that case.

Reviewer 4: Update looks good.

Reviewer 4: And whatever you're talking about regarding beanutils sounds like nobody noticed the version ranges specified in the Snyk page. The latest release, 1.9.4, is _not_ vulnerable to either of the linked CVEs. Being a library, its transitive dependencies are trivially overridden. If this dependency is being removed purely because a security scanner thinks a transitive dependency that is overridden elsewhere is still somehow a problem, then we might as well just boot all the real security engineers who work on this project since nobody seems to care what we think.

> It contains security vulnerabilities and also is not actually used by any part of Jenkins Core itself.

Reviewer 5: point taken @bitwiseman 

```suggestion
It is not actually used by any part of Jenkins Core itself.
It is also inherently risky from a security perspective:

* XML is overcomplicated and the XML parsing might be subject to XXE, maybe allowing clever attacks to do things like load `../../../credentials.xml`.
* Even if you make sure it is being fed a plain old self-contained XML document, you are still taking user-controlled strings and using them to look up and invoke Java methods, which is subject to abuse in numerous ways.
```

All but a few of the plugins affected by this change are some combination of unmaintained, rarely used, and no longer or never published.
We submitted PRs to all affected plugins.

=== Detached plugin to maintain compatibility

The current plan causes plugins which depend on Jenkins to provide Digester to fail unless they are updated.
This could be mitigated by moving this dependency to a detached plugin.
We decided against creating a detached pluging because there were a small number of affected plugins and only a few of them have significant install base.
The creating and maintaining of a detached plugin would still be a significant amount of work and would cause the security vulnerabilities we are trying to address to remain open.

== Backwards Compatibility

The current plan causes plugins which depend on Jenkins to provide Digester to fail unless they are updated.
The majority of this JEP is devoted to discussion of how to mitigate these breaks.

== Security

The motivation for this change is mitigating a number of known security issues

== Infrastructure Requirements

There are no new infrastructure requirements related to this proposal.

== Testing

> There are no testing issues related to this proposal.

Reviewer 5: `plugin-compat-tester` will need to be run on a reasonably large set of plugins to check for regressions introduced by the proposed core change.

Author: The proposed core change removes `commons-digester` and the `Digester2` API.  We've done a search for references to those in plugins in the `jenkinsci` and have filed PRs to address those.  What would you expect the `plugin-compat-tester` run to catch that a straight code search hasn't?

Reviewer 5: OK, if you are confident everything that might break would be caught by existing code searches, feel free to mark this & https://github.com/jenkinsci/jep/pull/361#discussion_r629703813 resolved.

Author: @jglick @timja 
I'm confident, but I'm willing to acknowledge I may be missing something.  

What would you expect the `plugin-compat-tester` run to catch that a straight code search hasn't?
What would it take to do a PCT run?

Regardless of the result of this conversation, I'll note this in the Reasoning section.

Reviewer 5: > What would you expect the `plugin-compat-tester` run to catch that a straight code search hasn't?

Nothing that I can think of in this case. In most core patches we would run PCT.

> What would it take to do a PCT run?

CloudBees has facilities to do so but the Jenkins project currently does not, alas.

| 18
|===

==== Affected plugins (no fixes planned)

===== Suspended plugins

Some plugins were found which were suspended when Jenkins 2 was released.
These will not be updated as part of this effort.

* https://plugins.jenkins.io/svn-release-mgr suspended since https://issues.jenkins-ci.org/browse/INFRA-2487
* https://github.com/jenkinsci/cpptest-plugin suspended since https://issues.jenkins-ci.org/browse/INFRA-2487
* https://github.com/jenkinsci/tfs-plugin suspended since https://issues.jenkins-ci.org/browse/INFRA-2751
* https://github.com/jenkinsci/cflint-plugin (link:https://github.com/jenkinsci/CFLint-plugin/pull/3[PR-3]) suspended since https://issues.jenkins-ci.org/browse/INFRA-2751
* https://github.com/jenkinsci/script-scm-plugin SECURITY-461
* https://github.com/jenkinsci/rtc-plugin (superseded by team-concert)

===== Never released plugins

Some plugins were never released.
These will not be updated as part of this effort.

* https://github.com/jenkinsci/cocoemma-plugin
* https://github.com/jenkinsci/jwsdp-sqe-plugin
* https://github.com/jenkinsci/pucm-plugin
* https://github.com/jenkinsci/purecm-plugin

== Motivation

> The link:https://commons.apache.org/proper/commons-digester/[Apache Commons Digester] that was included as a dependency of Jenkins Core is old and poorly maintained.
> It is not actually used by any part of Jenkins Core itself.
> It is also inherently risky from a security perspective:
> 
> * XML is overcomplicated and the XML parsing might be subject to XXE, maybe allowing clever attacks to do things like load `../../../credentials.xml`.
> * Even if you make sure it is being fed a plain old self-contained XML document, you are still taking user-controlled strings and using them to look up and invoke Java methods, which is subject to abuse in numerous ways.
> Jenkins Core exposes an API to plugin that mitigates some of these security issues, but it is not the job of Jenkins to do this.
> Instead, we should remove the Digester dependency from Jenkins Core and the plugins that actually use Digester should handle these security concerns themselves.
> Also, because it is a dependency of Core, some plugins have taken a dependency on Digester being provided by Jenkins rather than using the API provided by Jenkins.
> 
> Plugins in either of these scenarios will need to be updated or they will break after this removal.

Author: ```suggestion
The link:https://commons.apache.org/proper/commons-digester/[Apache Commons Digester]
included as a dependency of Jenkins Core is old and poorly maintained.
Digester and its dependencies have been a source of a number of security vulnerabilities. 
Examples of this include: 

* link:https://nvd.nist.gov/vuln/detail/CVE-2020-2304[CVE-2020-2304]
* link:https://nvd.nist.gov/vuln/detail/CVE-2020-2305[CVE-2020-2305]
* link:https://nvd.nist.gov/vuln/detail/CVE-2018-1000054[CVE-2018-1000054]
* link:https://nvd.nist.gov/vuln/detail/CVE-2018-1000055[CVE-2018-1000055]

The version of Digester could be updated, but in some sense any use of Digester is inherently risky from a security perspective:

* XML is overcomplicated and the XML parsing might be subject to XXE, allowing clever attacks to do things like load `../../../credentials.xml`.
* Even after verifying the input is a plain old self-contained XML document, Digester still takes user-controlled strings and uses them to look up and invoke Java methods, which is subject to abuse in numerous ways.

Over 15 years ago, the `Digester2` class was added to Jenkins Core to address 
link:https://issues.apache.org/jira/browse/DIGESTER-118[DIGESTER-118].
In the years since then, all usages of `Digester` and `Digester2` have been removed from Jenkins Core itself.
They remain only to avoid breaking plugins that depend on `Digester` and `Digester2` .

However, this means that addressing security issues related to `Digester` generally requires an update to Jenkins Core.
Also, because Digester is a dependency of Core, some plugins continue to use the `Digester` class being provided by from the dependency rather than using the `Digester2` API provided by Jenkins.

The right thing to do is to remove the Digester dependency from Jenkins Core along with the `Digester2` class.
Then the plugins that actually use Digester can handle any security concerns themselves.
Plugins that use `Digester` or `Digester2` will need to be updated or they will break after this removal, however only those plugins will need updating.
```

== Reasoning

=== Risks of complete removal

We chose to remove Digester from Jenkins Core in a manner that causes plugins which depend on it to fail unless they are updated.
We made this choice based on an analysis of all plugins which are currently distributed by the Jenkins Update Center or which are
in a repository in the link:https://github.com/jenkinsci[`jenkinsci` org on GitHub].
All but a few of the plugins affected by this change are some combination of unmaintained, rarely used, and no longer or never published.
We submitted PRs to all affected plugins.

=== Detached plugin to maintain compatibility

The current plan causes plugins which depend on Jenkins to provide Digester to fail unless they are updated.
This could be mitigated by moving this dependency to a detached plugin.
We decided against creating a detached pluging because there were a small number of affected plugins and only a few of them have significant install base.
The creating and maintaining of a detached plugin would still be a significant amount of work and would cause the security vulnerabilities we are trying to address to remain open.

> 

Author: ```suggestion

=== Testing limited to known plugins

We considered using `plugin-compat-tester` to test on a broad set of plugins beyond those known to be affected
(link:https://github.com/jenkinsci/jep/pull/361/files#r633849804[PR thread]). It was decided that the code searches we did within the `jenkinsci` repositories were sufficient. We could not think of any failures that would be caught by the PCT that were not already addressed by the code search. 
```
